cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+		zio_cmd->zv = zinfo;
+		taskq_dispatch(zinfo->uzfs_zvol_taskq, uzfs_zvol_worker,
+		    zio_cmd, TQ_SLEEP);
+	}
+exit:
+	if (zinfo != NULL) {
+
+		(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
+		zinfo->conn_closed = true;
+		zinfo->is_io_ack_sender_created = 0;
+		(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
+		/*
+		 * Send signal to ack sender so that it can free
+		 * zio_cmd, close fd and exit.
+		 */
+		(void) pthread_mutex_lock(&zinfo->complete_queue_mutex);
+		if (zinfo->io_ack_waiting) {
+			rc = pthread_cond_signal(&zinfo->io_ack_cond);
+		}
+		(void) pthread_mutex_unlock(&zinfo->complete_queue_mutex);
+		uzfs_zinfo_drop_refcnt(zinfo, false);
+	}
+
+	printf("uzfs_zvol_io_receiver thread exiting\n");
+	zk_thread_exit();
+}
+
+/*
+ * This function suppose to lookup into zvol list
+ * to find if LUN presented for identification is
+ * available/online or not. This function also need
+ * to return IP address of replica along with port
+ * so that ISTGT controller can open a connection
+ * for IOs.
+ */
+static int
+uzfs_zvol_mgmt_do_handshake(zvol_io_hdr_t *hdr, int sfd, char *name)
+{
+	int 		count, rc = 0;
+	zvol_info_t 	*zinfo;
+	mgmt_ack_t 	mgmt_ack = {0, };
+
+	char 		*packet = NULL;
+	char 		*p = NULL;
+
+	printf("Volume: %s sent for enq\n", name);
+	hdr->len = sizeof (mgmt_ack_t);
+	zinfo = uzfs_zinfo_lookup(name);
+	if (zinfo != NULL) {
+		hdr->status = ZVOL_OP_STATUS_OK;
+	} else {
+		hdr->status = ZVOL_OP_STATUS_FAILED;
+	}
+
+	strncpy(mgmt_ack.volname, name, strlen(name));
+	mgmt_ack.port = 3232;
+	rc = uzfs_zvol_get_ip(mgmt_ack.ip);
+	printf("IP address: %s\n", mgmt_ack.ip);
+	if (rc == -1) {
+		ZREPL_ERRLOG("Unable to get IP"
+		    " with err:%d\n", errno);
+		goto exit;
+	}
+
+	packet = kmem_alloc((sizeof (mgmt_ack_t) + sizeof (*hdr)) *
+	    sizeof (char), KM_SLEEP);
+	bcopy(hdr, packet, sizeof (*hdr));
+	p = packet + sizeof (*hdr);
+	bcopy(&mgmt_ack, p, sizeof (mgmt_ack));
+	count = write(sfd, packet, (sizeof (*hdr) + sizeof (mgmt_ack_t)));
+	if (count == -1) {
+		ZREPL_ERRLOG("Write to socket failed"
+		    " with err:%d\n", errno);
+		rc = -1;
+		goto exit;
+	}
+exit:
+	uzfs_zinfo_drop_refcnt(zinfo, false);
+	return (rc);
+}
+
+/*
+ * TODO: IMHO, this should be a through away API.
+ * Side Car has to find a more smart way to pass
+ * ISCSI Controller IP address.
+ */
+static char *
+get_controller_ip_address(void)
+{
+	size_t nbytes;
+	char *buf = NULL;
+
+	FILE *fp = fopen("/var/openebs/controllers.conf", "r");
+	if (fp == NULL) {
+		printf("Error opening file\n");
+		return (buf);
+	}
+
+	buf = kmem_alloc(1024, KM_SLEEP);
+	nbytes = fread(buf, sizeof (char), 1024, fp);
+
+	if (nbytes == 0) {
+		printf("Read error\n");
+		return (buf);
+	}
+	printf("buffer value is %s", buf);
+	return (buf);
+}
+
+/*
+ * One thread per replica, which will be
+ * responsible for initial handshake and
+ * exchanging info like IP add, port etc.
+ */
+static void
+uzfs_zvol_mgmt_thread(void *arg)
```

Comment:
do I understand correctly that as time will go we will add more management commands on this connection? Or is the handshake command the only one?